### PR TITLE
[DRAFT][quantization] Full quantization of LLama compatible models

### PR DESCRIPTION
**Pull request by stamalakhov**
This draft tries to get fully quantized circle layers for `Llama` model.

TODO:
- tests/cleanup

TICO-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

### CODE REVIEW

File: tico/quantization/algorithm/gptq/quantizer.py [hunk 1]
```diff
@@ -212,7 +212,7 @@ def convert(self, model):
                 for name in subset:
                     gptq[name] = GPTQ(subset[name])
                     gptq[name].quantizer.configure(
-                        bits=8, perchannel=True, sym=False, mse=False
+                        bits=4, perchannel=True, sym=False, mse=False
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/algorithm/gptq/quantizer.py]: > FYI, you can give the option for this with [this PR](https://github.com/Samsung/TICO/pull/441).

@mhs4670go 
Thank you. I'll rebase after merging of #441.

File: tico/quantization/wrapq/observers/affine_base.py
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/observers/affine_base.py]: > Could you give me some explanations for the reason of changes related with observers?
>1. Deleting some attributes and register them as buffer.

Ahh. It occured that `model.to("cuda")` or `model.to("cpu")` do not transfer `scales` and `zero_points` to `gpu`/`cpu`, they were not registered as `buffers` or `parameters`, that is why they were registered as `buffers`. Deleting them is needed, because otherwise `torch` fails to register known `attributes` as `buffers`.

>2. Change ObserverBase's parent from ABC to torch.nn.Module

It will enable using buffer `registering` and correct automatical transfer of `scales`/`zp` to `cpu`<->`gpu`. The same approach is used in `gptq/quant.py` for the same reason (i suppose).
Please see https://github.com/Samsung/TICO/blob/aaf55d782861eb8c9232d11329c4105a08db24fb/tico/quantization/algorithm/gptq/quant.py#L32

This draft is just PoC (quick and dirty).

File: tico/quantization/passes/fold_quant_ops.py [hunk 2]
```diff
@@ -144,7 +144,78 @@ def call(self, exported_program: ExportedProgram) -> PassResult:
                     )
                     dq.replace_all_uses_with(op, propagate_meta=False)
                     logger.debug(f"Removed redundant {dq.name}")
+            
+        for dq in graph.nodes:
+            if dq.op != "call_function":
+                continue
+            if (
+                dq.target
+                != torch.ops.circle_custom.dequantize_mx_to_float.default
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/passes/fold_quant_ops.py]: > Ah, I see. How about go with `quantize_mx_decomposed`, and `dequantize_mx_decomposed`? This aligns with `torch.ops.quantized_decomposed`.

@mhs4670go 
Ok. Got it. Thank you.

File: tico/quantization/wrapq/observers/affine_base.py
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/observers/affine_base.py]: > Ah, thanks for the clarification. I'll reconsider those and apply the changes soon.

ok. Thank you very much.

File: tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+        attn_mask = torch.ones_like(rnd_input)
+        position_ids = torch.arange(q_m.config.max_position_embeddings).unsqueeze(0)
+        device = "cuda"
+        exported_program = torch.export.export(
+            q_m.to(device), (rnd_input.to(device), attn_mask.to(device), position_ids.to(device)), kwargs=None, dynamic_shapes=None, strict=False
+        )
+        ppl = evaluate_ppl_of_exported_module_on_dataset(exported_program.module(), test_ids_batch, device=device)
+        print("\n┌── Wikitext-2 test perplexity ─────────────")
+        print(f"│ exported_{args.default_io_qdtype} : {ppl:8.2f}")
+        print("└───────────────────────────────────────────")
```
Comment (stamalakhov) [inline tico/quantization/wrapq/examples/quantize_full_qmodel_with_gptq.py]: @mhs4670go It would be nice to provide `ppl` for the final `circle` model, but as far as i know `onert` doesn't use comparable with `cuda` accelerated kernels. May be it is possible to compute `ppl` right before `circle` serialization (right after all passes), so that serialization is very unlikely to break the model? 
I mean that it's possible to provide needed custom quantized kernels to pytorch, like `quantized_matmul` to mimick quantized computations in `circle-interpreter`.
Please correct me if i'm wrong.

File: tico/quantization/wrapq/wrappers/nn/quant_embedding.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from tico.quantization.config.ptq import PTQConfig
+
+from tico.quantization.wrapq.mode import Mode
+from tico.quantization.wrapq.qscheme import QScheme
+from tico.quantization.wrapq.wrappers.quant_module_base import QuantModuleBase
+from tico.quantization.wrapq.wrappers.registry import try_register
+
+
+@try_register("torch.nn.Embedding")
+class QuantEmbedding(QuantModuleBase):
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/nn/quant_embedding.py]: Ok. Thank you!

File: tico/quantization/wrapq/wrappers/llama/quant_attn.py [hunk 5]
```diff
@@ -208,10 +210,15 @@ def forward(
...
         # HF Cache path (if available)
-        if use_cache and hasattr(past_key_value, "update"):
+        if (use_cache or self.return_kv_cache) and hasattr(past_key_value, "update"):
+
+            sig = signature(past_key_value.update)
             # Many HF Cache impls use update(k, v) and return (k_total, v_total)
             try:
-                k_total, v_total = past_key_value.update(k_rot, v)
+                if len(sig.parameters) == 2:
+                    k_total, v_total = past_key_value.update(k_rot, v)
```
[the remark this answers is not in the document]
Comment (stamalakhov) [reply tico/quantization/wrapq/wrappers/llama/quant_attn.py]: `if len(sig.parameters) == 2` to make test passable

File: tico/quantization/wrapq/quantizer.py [hunk 6]
```diff
@@ -84,6 +84,45 @@ def _wrap_supported(
 
         # Case A: HuggingFace-style transformers: model.model.layers
         lm = getattr(root, "model", None)
+
+        embeddings = (
```
Comment (stamalakhov) [inline tico/quantization/wrapq/quantizer.py]: @mhs4670go 
May be it will be better to introduce something like `QuantLlamaModel` and wrap all internal structure (`embed`, `lm_head`, `norm`) inside of it? and leave this general code as a fallback? like this:
```
try:
    wrap(the_whole_model)
catch:
  no specific wrapper for a class then use  general logic
```

Comment (mhs4670go) [reply tico/quantization/wrapq/quantizer.py]: It would be good to have `QuantLlamaModel` for convinience when we evaluate it or something like that. Just for note, even though the whole model is quantized, only decoder layers would need to be exported because of the runtime requirements. Nothing has been fixed though.

Comment (stamalakhov) [reply tico/quantization/wrapq/quantizer.py]: OK. Got it. Thank you.

**Comment by stamalakhov** [comment]
@mhs4670go
I've updated this draft to use mx types for quantizing activations for `rms_norm`, `softmax`, `matmuls`.
Do we need `MXTypes` in `main` branch?
